serpent-moltres/full-core/extract-convert.py

```diff
@@ -0,0 +1,320 @@
+#!/usr/bin/env python3
+# This file is based in $MOLTRES/python/extractSerpent2GCs.py
+import os
+import numpy as np
+import argparse
+import subprocess
+import serpentTools as sT
+
+
+def makePropertiesDir(outdir, filebase, mapFile, unimapFile, ngroups='3'):
+    '''
+    Takes in a mapping from universe names to material properties.
+    It collapses the material properties from 26 to ngroups.
+    Then, it makes a material properties directory.
+
+    Parameters:
+    -----------
+    outdir: [string]
+        name of directory to write properties to
+    filebase: [string]
+        file base name to give moltres
+    mapFile: [string]
+        file that maps branches to temperatures
+    unimapFile: [string]
+        maps material names to serpent universe
+    ngroups: [string]
+        the options are 3, 6, 9, 12, 15, 15b, 15c, 15d, 15e, 18, 18c, 21
+    '''
+
+    # the constants moltres looks for:
+    goodStuff = ['BETA_EFF', 'Chit', 'Chid', 'lambda', 'Diffcoef', 'Kappa',
+                 'Sp0', 'Nsf', 'Invv', 'Remxs', 'Fiss', 'Flx', 'Abs']
+    goodMap = dict([(thing, 'inf' + thing) for thing in goodStuff])
+    goodMap['BETA_EFF'] = 'betaEff'
+    goodMap['lambda'] = 'lambda'
+
+    # map material names to universe names from serpent
+    with open(unimapFile) as fh:
+        uniMap = []
+        for line in fh:
+            uniMap.append(tuple(line.split()))
+    # this now maps material names to serpent universes
+    uniMap = dict(uniMap)
+
+    # list of material names
+    inmats = list(uniMap.keys())
+
+    print("Making properties for materials:")
+    print(inmats)
+    coeList = dict([(mat, sT.read(mat + '.coe')) for mat in inmats])
+
+    # primary branch to temp mapping
+    branch2TempMapping = open(mapFile)
+
+    # Check if calculation uses 6 neutron precursor groups.
+    # This prevents writing of excess zeros. Check if any
+    # entries in the 7th and 8th group precursor positions
+    # are nonzero, if so, use 8 groups.
+    use8Groups = False
+    for line in branch2TempMapping:
+        item, temp = tuple(line.split())
+        for mat in inmats:
+            if mat in item:
+                currentMat = mat
+                break
+
+        strData = coeList[currentMat].branches[item].universes[
+            uniMap[currentMat], 0, 0, None].gc[goodMap['BETA_EFF']]
+        strData = strData[1:9]
+        if np.any(strData[-2:] != 0.0):
+            use8Groups = True
+
+    # Now loop through a second time
+    branch2TempMapping.close()
+    branch2TempMapping = open(mapFile)
+
+    for line in branch2TempMapping:
+
+        item, temp = tuple(line.split())
+        for mat in inmats:
+            if mat in item:
+                currentMat = mat
+                break
+        else:
+            print('Considered materials: {}'.format(inmats))
+            raise Exception(
+                'Couldnt find a material corresponding to branch {}'.format(
+                    item))
+
+        try:
```

Comment:
This try block is really something.... I don't like it.  Can you give these group structures meaningful names rather than repeatedly hardcoding the list?